Assertion error related to ndarray/extract

to replicate:

```python
phenotypes_BP = ['is_BP1', 'is_BP2', 'is_BP', 'is_BP_including_BPSCZ', 'is_BPNOS', 'is_BPSCZ', 'is_BPPSY', 'is_BP_no_PSY']
def make_fake_data():
    mt = hl.utils.range_matrix_table(1000, 100)
    mt = mt.annotate_rows(gene_set = "my_gene", consequence_category = "cnsq")
    mt = mt.annotate_cols(phenotype_boolean = hl.struct(**{pheno : hl.rand_bool(.5) for pheno in phenotypes_BP}))
    mt = mt.annotate_cols(is_female = hl.rand_bool(.5))
    mt = mt.annotate_cols(col_id = hl.str(mt.col_idx))
    mt = mt.annotate_cols(**{f"PC{i + 1}":hl.rand_unif(0, 10) for i in range(10)})
    mt = mt.annotate_entries(x = hl.rand_unif(0, 3))
    mt = mt.key_rows_by(mt.gene_set, mt.consequence_category)
    mt = mt.key_cols_by(mt.col_id)
    return mt

def permute_phenotypes(np_pheno, n_perms):

    np_pheno[np_pheno == None] = 2
    np_pheno = np_pheno.astype(int)
    np_pheno_mat = np.repeat(np_pheno, n_perms).reshape(np_pheno.size, n_perms)
    for i in range(np_pheno_mat.shape[1]):
        np.random.shuffle(np_pheno_mat[:,i])

    return(np_pheno_mat)

def run_regressions_perm(mt, phenotypes, n_perms):

    mt = mt.add_col_index()

    init = True
    for phenotype in phenotypes:

        mt = mt.annotate_globals(
            pheno_perm = permute_phenotypes(np.array(mt.phenotype_boolean[phenotype].collect()), n_perms))
        mt = mt.annotate_globals(
            pheno_perm = mt.pheno_perm.map(
                lambda x: { hl.array([False, True, hl.null(hl.tbool)])[hl.int(x)] }
            ))
        return mt

mt = make_fake_data()
res = run_regressions_perm(mt, phenotypes_BP, 100)
res.show()
```

error:

```
java.lang.AssertionError: assertion failed
    at scala.Predef$.assert(Predef.scala:156)
    at is.hail.expr.ir.agg.Extract$.extract(Extract.scala:331)
    at is.hail.expr.ir.agg.Extract$.is$hail$expr$ir$agg$Extract$$extract$1(Extract.scala:215)
    at is.hail.expr.ir.agg.Extract$$anonfun$extract$3.apply(Extract.scala:333)
    at is.hail.expr.ir.agg.Extract$$anonfun$extract$3.apply(Extract.scala:333)
    at is.hail.expr.ir.MapIR$$anonfun$apply$1.apply(MapIR.scala:8)
    at is.hail.expr.ir.MapIR$$anonfun$apply$1.apply(MapIR.scala:7)
    at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
    at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
    at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:35)
    at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
    at scala.collection.AbstractTraversable.map(Traversable.scala:104)
    at is.hail.expr.ir.MapIR$.apply(MapIR.scala:7)
    at is.hail.expr.ir.agg.Extract$.extract(Extract.scala:333)
    at is.hail.expr.ir.agg.Extract$.is$hail$expr$ir$agg$Extract$$extract$1(Extract.scala:215)
    at is.hail.expr.ir.agg.Extract$$anonfun$extract$3.apply(Extract.scala:333)
    at is.hail.expr.ir.agg.Extract$$anonfun$extract$3.apply(Extract.scala:333)
    at is.hail.expr.ir.MapIR$$anonfun$apply$1.apply(MapIR.scala:8)
    at is.hail.expr.ir.MapIR$$anonfun$apply$1.apply(MapIR.scala:7)
    at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
    at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
    at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:35)
    at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
    at scala.collection.AbstractTraversable.map(Traversable.scala:104)
    at is.hail.expr.ir.MapIR$.apply(MapIR.scala:7)
    at is.hail.expr.ir.agg.Extract$.extract(Extract.scala:333)
    at is.hail.expr.ir.agg.Extract$.is$hail$expr$ir$agg$Extract$$extract$1(Extract.scala:215)
    at is.hail.expr.ir.agg.Extract$$anonfun$extract$3.apply(Extract.scala:333)
    at is.hail.expr.ir.agg.Extract$$anonfun$extract$3.apply(Extract.scala:333)
    at is.hail.expr.ir.MapIR$$anonfun$apply$1.apply(MapIR.scala:8)
    at is.hail.expr.ir.MapIR$$anonfun$apply$1.apply(MapIR.scala:7)
    at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
    at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
    at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:35)
    at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
    at scala.collection.AbstractTraversable.map(Traversable.scala:104)
    at is.hail.expr.ir.MapIR$.apply(MapIR.scala:7)
    at is.hail.expr.ir.agg.Extract$.extract(Extract.scala:333)
    at is.hail.expr.ir.agg.Extract$.is$hail$expr$ir$agg$Extract$$extract$1(Extract.scala:215)
    at is.hail.expr.ir.agg.Extract$$anonfun$extract$3.apply(Extract.scala:333)
    at is.hail.expr.ir.agg.Extract$$anonfun$extract$3.apply(Extract.scala:333)
    at is.hail.expr.ir.MapIR$$anonfun$apply$1.apply(MapIR.scala:8)
    at is.hail.expr.ir.MapIR$$anonfun$apply$1.apply(MapIR.scala:7)
    at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
    at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
    at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:35)
    at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
    at scala.collection.AbstractTraversable.map(Traversable.scala:104)
```

--- a/hail/python/hail/expr/expressions/base_expression.py
+++ b/hail/python/hail/expr/expressions/base_expression.py
@@ -724,7 +724,7 @@ class Expression(object):
             raise NotImplementedError('cannot convert aggregated expression to table')
 
         if source is None:
-            return fallback_name, Env.dummy_table().select(**{fallback_name: self})
+            return fallback_name, hl.Table.parallelize([self], n_partitions=1)
 
         name = source._fields_inverse.get(self)
         top_level = name is not None
@@ -733,7 +733,7 @@ class Expression(object):
         named_self = {name: self}
         if len(axes) == 0:
             x = source.select_globals(**named_self)
-            ds = Env.dummy_table().select(**{name: x.index_globals()[name]})
+            ds = hl.Table.parallelize([x.index_globals()], n_partitions=1)
         elif isinstance(source, hail.Table):
             if top_level and name in source.key:
                 named_self = {}

--- a/hail/python/hail/expr/expressions/base_expression.py
+++ b/hail/python/hail/expr/expressions/base_expression.py
@@ -724,7 +724,7 @@ class Expression(object):
             raise NotImplementedError('cannot convert aggregated expression to table')
 
         if source is None:
-            return fallback_name, hl.Table.parallelize([self], n_partitions=1)
+            return fallback_name, hl.Table.parallelize([hl.struct(**{fallback_name: self})], n_partitions=1)
 
         name = source._fields_inverse.get(self)
         top_level = name is not None

--- a/hail/python/test/hail/matrixtable/test_matrix_table.py
+++ b/hail/python/test/hail/matrixtable/test_matrix_table.py
@@ -1531,3 +1531,10 @@ class Tests(unittest.TestCase):
         hl.experimental.write_matrix_tables(mts, f)
         assert hl.read_matrix_table(f + '0.mt')._same(mts[0])
         assert hl.read_matrix_table(f + '1.mt')._same(mts[1])
+
+    def test_key_cols_by_extract_issue(self):
+        mt = hl.utils.range_matrix_table(1000, 100)
+        mt = mt.key_cols_by(col_id = hl.str(mt.col_idx))
+        mt = mt.add_col_index()
+        mt.show()
+

--- a/hail/src/main/scala/is/hail/expr/ir/Env.scala
+++ b/hail/src/main/scala/is/hail/expr/ir/Env.scala
@@ -24,11 +24,11 @@ case class BindingEnv[V](
   def allEmpty: Boolean = eval.isEmpty && agg.forall(_.isEmpty) && scan.forall(_.isEmpty)
 
   def promoteAgg: BindingEnv[V] = {
-    BindingEnv(agg.get)
+    BindingEnv(agg.get, scan = scan)
   }
 
   def promoteScan: BindingEnv[V] = {
-    BindingEnv(scan.get)
+    BindingEnv(scan.get, agg = agg)
   }
 
   def bindEval(name: String, v: V): BindingEnv[V] =

--- a/hail/src/main/scala/is/hail/expr/ir/Exists.scala
+++ b/hail/src/main/scala/is/hail/expr/ir/Exists.scala
@@ -62,7 +62,7 @@ object ContainsAgg {
     case l: AggLet => !l.isScan
     case _: TableAggregate => false
     case _: MatrixAggregate => false
-    case _: StreamAgg => true // this should be permitted, but causes problems elsewhere in the IR
+    case _: StreamAgg => false
     case _ => root.children.exists {
       case child: IR => ContainsAgg(child)
       case _ => false
@@ -111,7 +111,7 @@ object ContainsScan {
     case l: AggLet => l.isScan
     case _: TableAggregate => false
     case _: MatrixAggregate => false
-    case _: StreamAggScan => true // this should be permitted, but causes problems elsewhere in the IR
+    case _: StreamAggScan => false
     case _ => root.children.exists {
       case child: IR => ContainsScan(child)
       case _ => false

--- a/hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala
+++ b/hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala
@@ -286,10 +286,8 @@ object TypeCheck {
         assert(body.typ == TVoid)
       case x@StreamAgg(a, name, query) =>
         assert(a.typ.isInstanceOf[TStream])
-        assert(env.agg.isEmpty)
       case x@StreamAggScan(a, name, query) =>
         assert(a.typ.isInstanceOf[TStream])
-        assert(env.scan.isEmpty)
         assert(x.typ.asInstanceOf[TStream].elementType == query.typ)
       case x@RunAgg(body, result, _) =>
         assert(x.typ == result.typ)

--- a/hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala
+++ b/hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala
@@ -114,6 +114,7 @@ case class Aggs(postAggIR: IR, init: IR, seqPerElt: IR, aggs: Array[AggStateSign
 object Extract {
   def liftScan(ir: IR): IR = ir match {
     case ApplyScanOp(init, seq, sig) => ApplyAggOp(init, seq, sig)
+    case StreamAggScan(a, name, query) => StreamAggScan(liftScan(a), name, query)
     case x => MapIR(liftScan)(x)
   }
 

--- a/hail/src/main/scala/is/hail/lir/X.scala
+++ b/hail/src/main/scala/is/hail/lir/X.scala
@@ -64,7 +64,7 @@ class Classx[C](val name: String, val superName: String) {
           else {
             /*
             if (l.method ne m) {
-              println(s"$l ${l.method} $m")
+              println(s"$l ${l.method} $m\n  ${l.stack.mkString("  \n")}")
             }
              */
             assert(l.method eq m)
@@ -317,6 +317,7 @@ class MethodLit(
 
 class Local(var method: Method, val name: String, val ti: TypeInfo[_]) {
   override def toString: String = f"t${ System.identityHashCode(this) }%08x/$name"
+//  val stack = Thread.currentThread().getStackTrace
 }
 
 class Parameter(method: Method, val i: Int, ti: TypeInfo[_]) extends Local(method, null, ti) {

--- a/hail/src/test/scala/is/hail/expr/ir/TableIRSuite.scala
+++ b/hail/src/test/scala/is/hail/expr/ir/TableIRSuite.scala
@@ -431,4 +431,50 @@ class TableIRSuite extends HailSuite {
     val tir = TableRead(tr.fullType, true, tr)
     assert(tir.partitionCounts.forall(_.sum == 0))
   }
+
+  @Test def testScanInAggInMapRows() {
+    implicit val execStrats = ExecStrategy.interpretOnly
+    val sumSig = AggSignature(Sum(), FastSeq(), FastSeq(TInt64))
+    var tr: TableIR = TableRange(10, 3)
+    tr = TableKeyBy(tr, FastIndexedSeq(), false)
+    tr = TableMapRows(tr, InsertFields(Ref("row", tr.typ.rowType),
+      FastSeq(("result",
+        StreamAgg(
+          StreamAggScan(
+            StreamRange(0, GetField(Ref("row", tr.typ.rowType), "idx"), 1),
+            "streamx",
+            ApplyScanOp(FastIndexedSeq(), FastIndexedSeq(Ref("streamx", TInt32).toL), sumSig)),
+          "aggx",
+          ApplyAggOp(FastIndexedSeq(), FastIndexedSeq(Ref("aggx", TInt64)), sumSig))))))
+    assertEvalsTo(TableCollect(tr), Row(IndexedSeq.tabulate(10) { i =>
+      val r = (0 until i).map(_.toLong).scanLeft(0L)(_ + _).init.sum
+      Row(i, r)
+    }, Row()
+    ))
+  }
+
+  @Test def testScanInAggInScanInMapRows() {
+    implicit val execStrats = ExecStrategy.interpretOnly
+    val sumSig = AggSignature(Sum(), FastSeq(), FastSeq(TInt64))
+    var tr: TableIR = TableRange(10, 3)
+    tr = TableKeyBy(tr, FastIndexedSeq(), false)
+    tr = TableMapRows(tr, InsertFields(Ref("row", tr.typ.rowType),
+      FastSeq(("result",
+        ApplyScanOp(FastIndexedSeq(),
+          FastIndexedSeq(StreamAgg(
+            StreamAggScan(
+              StreamRange(0, GetField(Ref("row", tr.typ.rowType), "idx"), 1),
+              "streamx",
+              ApplyScanOp(FastIndexedSeq(), FastIndexedSeq(Ref("streamx", TInt32).toL), sumSig)),
+            "aggx",
+            ApplyAggOp(FastIndexedSeq(), FastIndexedSeq(Ref("aggx", TInt64)), sumSig))),
+          sumSig)))))
+    assertEvalsTo(TableCollect(tr), Row(Array.tabulate(10) { i =>
+      (0 until i).map(_.toLong).scanLeft(0L)(_ + _).init.sum
+    }.scanLeft(0L)(_ + _)
+      .zipWithIndex
+      .map { case (x, idx) => Row(idx, x) }.init.toFastIndexedSeq,
+      Row()
+    ))
+  }
 }

